src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/AbstractConverter.cs

```diff
@@ -0,0 +1,78 @@
+
+        public AbstractConverter(ForEachInfo<ForEachStatementSyntax, StatementSyntax> forEachInfo)
+        {
+            ForEachInfo = forEachInfo;
+        }
+
+        /// <summary>
+        /// Creates a query expression.
+        /// </summary>
+        /// <param name="selectExpression">expression to be used into the last 'select ...' in the query expression</param>
+        /// <param name="leadingTokenForSelect">extra leading tokens to be added to the select clause</param>
+        /// <param name="trailingTokenForSelect">extra trailing tokens to be added to the select clause</param>
+        /// <returns></returns>
+        protected QueryExpressionSyntax CreateQueryExpression(
+            ExpressionSyntax selectExpression,
+            IEnumerable<SyntaxToken> leadingTokenForSelect,
+            IEnumerable<SyntaxToken> trailingTokenForSelect)
+            => SyntaxFactory.QueryExpression(
+                CreateFromClause(ForEachInfo.ForEachStatement, Helpers.GetTrivia(ForEachInfo.LeadingTokens), Enumerable.Empty<SyntaxTrivia>()),
+                SyntaxFactory.QueryBody(
+                    SyntaxFactory.List(ForEachInfo.ConvertingExtendedNodes.Select(node => CreateQueryClause(node))),
+                    SyntaxFactory.SelectClause(selectExpression)
+                    .WithComments(leadingTokenForSelect, ForEachInfo.TrailingTokens.Concat(trailingTokenForSelect)),
+                    continuation: null)) // The current coverage of foreach statements to support does not need to use query continuations.                                                                                                           
+            .WithAdditionalAnnotations(Formatter.Annotation);
+
+        private static QueryClauseSyntax CreateQueryClause(ExtendedSyntaxNode node)
+        {
+            switch (node.Node.Kind())
+            {
+                case SyntaxKind.VariableDeclarator:
+                    var variable = (VariableDeclaratorSyntax)node.Node;
+                    return SyntaxFactory.LetClause(SyntaxFactory.Token(SyntaxKind.LetKeyword), variable.Identifier, variable.Initializer.EqualsToken, variable.Initializer.Value)
+                        .WithComments(node.ExtraLeadingComments, node.ExtraTrailingComments);
+
+                case SyntaxKind.ForEachStatement:
+                    return CreateFromClause((ForEachStatementSyntax)node.Node, node.ExtraLeadingComments, node.ExtraTrailingComments);
+
+                case SyntaxKind.IfStatement:
+                    var ifStatement = (IfStatementSyntax)node.Node;
+                    return SyntaxFactory.WhereClause(
+                                SyntaxFactory.Token(SyntaxKind.WhereKeyword).WithComments(ifStatement.IfKeyword.LeadingTrivia, ifStatement.IfKeyword.TrailingTrivia),
+                                ifStatement.Condition.WithComments(new[] { ifStatement.OpenParenToken }, new[] { ifStatement.CloseParenToken }))
```

Comment:
1. Consider having an overload that just allows a leading/trailing token. 
2. Name should likely be WithCommentsFrom.